community/values/src/main/java/org/neo4j/values/virtual/NodeValue.java

```diff
@@ -68,4 +63,107 @@ public String toString()
         return format( "(%d)", id );
     }
 
+    static class DirectNodeValue extends NodeValue
+    {
+        private final TextArray labels;
+        private final MapValue properties;
+
+        DirectNodeValue( long id, TextArray labels, MapValue properties )
+        {
+            super( id );
+            assert labels != null;
+            assert properties != null;
+            this.labels = labels;
+            this.properties = properties;
+        }
+
+        public TextArray labels()
+        {
+            return labels;
+        }
+
+        public MapValue properties()
+        {
+            return properties;
+        }
+    }
+
+    public static class NodeProxyWrappingNodeValue extends NodeValue
+    {
+        private final Node node;
+        private volatile TextArray labels;
+        private volatile MapValue properties;
+
+        NodeProxyWrappingNodeValue( Node node )
+        {
+            super( node.getId() );
+            this.node = node;
+        }
+
+        public Node nodeProxy()
+        {
+            return node;
+        }
+
+        @Override
+        public <E extends Exception> void writeTo( AnyValueWriter<E> writer ) throws E
+        {
+            TextArray l;
+            MapValue p;
+            try
+            {
+                l = labels();
+                p = properties();
+            }
+            catch ( NotFoundException e )
+            {
+                l = Values.stringArray();
+                p = VirtualValues.EMPTY_MAP;
+
+            }
+            writer.writeNode( node.getId(), l, p );
+        }
+
+        @Override
+        public TextArray labels()
+        {
+            TextArray l = labels;
+            if ( l == null )
+            {
+                synchronized ( this )
```

Comment:
My thinking is that we do create a lot of `Node` and `Relationship` - and hence now `NodeValue` and `EdgeValue` - in the interpreted runtime. Most of these object are not actually returned to the user and thus eagerly pulling in labels, types and properties eagerly would add a big cost for all these intermediate objects. In this way the labels and properties will only be fetched when bolt actually tries to serialize the objects.

In a new glorious runtime I think we should be a lot more careful about creating these intermediate objects, but I felt addressing all the cases where we create nodes and relationships unnecessarily in this PR would of had serious impact on my motivation for finishing this piece of work :)